Fix for ResponseMetadata race condition.

zzmao opened the pull request:
Before this change, responseMetadata is accessed without confirming if it is ready or not.
This fix introduces a latch to avoid such race condition.

zzmao commented:
Internal Ticket: EXC-38958

zzmao commented:
Need to figure out why this happen.

zzmao commented:
Couldn't figure out where is the concurrency. GetBlobOperation::maybeWriteToChannel is the caller, but it uses while loop instead of multi thread. @cgtz  any insight?

cgtz commented:
I will look through the code a little more to see why this may happen. I worry that the latch may never get counted down if there is some other bug.

justinlin-linkedin commented:
> I will look through the code a little more to see why this may happen. I worry that the latch may never get counted down if there is some other bug. I guess the NPE could happen if there are two calls to `maybeWriteToChannel` in close proximity (concurrent write() calls?)

write call should be a blocking call that creates the finalRespoonseMetadata and maybeWriteToChannel seems like it's only invoked in a single thread.
Is it possible that the http channel is not active when we start the first write? in this case the finalResponseMetadata will not be constructed but yet we expect it in the Chunk constructor. Just. a thought.

zzmao commented:
I have same guess with @justinlin-linkedin ,

```
 if (responseMetadataWriteInitiated.compareAndSet(false, true) && ctx.channel().isActive()) {
```

It sets responseMetadataWriteInitiated to true, but channel is not active. 
@cgtz

zzmao commented:
Made another change. Please take a second look at. @cgtz @justinlin-linkedin

justinlin-linkedin commented on ambry-rest/src/main/java/com.github.ambry.rest/NettyResponseChannel.java:
```diff
@@ -148,6 +148,18 @@ public void onCompletion(Long result, Exception exception) {
     if (!responseMetadataWriteInitiated.get()) {
       maybeWriteResponseMetadata(responseMetadata, new ResponseMetadataWriteListener());
     }
+    if (finalResponseMetadata == null) {
```
this looks good, but can you add a test case for this logic?

cgtz replied:
+1, I guess we can at least test the case where the channel becomes inactive

justinlin-linkedin commented on ambry-rest/src/main/java/com.github.ambry.rest/NettyResponseChannel.java:
```diff
@@ -148,6 +148,18 @@ public void onCompletion(Long result, Exception exception) {
+        writeFuture.addListener(new CleanupCallback(new ClosedChannelException()));
+      }
+      FutureResult<Long> future = new FutureResult<Long>();
+      future.done(-1L, null);
```
the exception should be ClosedChannelException, not null.

cgtz replied:
Also call the callback here?

cgtz commented on ambry-rest/src/main/java/com.github.ambry.rest/NettyResponseChannel.java:
```diff
@@ -148,6 +148,18 @@ public void onCompletion(Long result, Exception exception) {
+    if (finalResponseMetadata == null) {
+      // If finalResponseMetadata is still null, it indicates channel becomes inactive.
+      if (ctx.channel().isActive()) {
+        logger.warn("Channel should be inactive status. {}", ctx.channel());
```
Could you add a metric for this? We can set up an alert if this case ever occurs.

zzmao commented:
@cgtz  @justinlin-linkedin 

Addressed comments and added test. Please take another round of review. Thanks!

The new test can reproduce the issue without fix and test passes with the fix.

cgtz commented on ambry-rest/src/test/java/com.github.ambry.rest/NettyResponseChannelTest.java:
```diff
@@ -1713,3 +1746,221 @@ void compareWithFuture() throws InterruptedException, TimeoutException {
 }
+
+/**
+ * Mock class for ChannelHandlerContext used in .
```
add "in channelInactiveTest()"

justinlin-linkedin commented on ambry-rest/src/test/java/com.github.ambry.rest/NettyResponseChannelTest.java:
```diff
@@ -636,6 +647,28 @@ public void closeTest() {
+    assertFalse("Channel should be closed", channel.isOpen());
+    Future<Long> future = nettyResponseChannel.write(Unpooled.buffer(1), null);
+    try {
+      future.get();
```
please add a fail("") after future.get();

cgtz commented on ambry-rest/src/main/java/com.github.ambry.rest/NettyMetrics.java:
```diff
@@ -170,6 +171,8 @@ public NettyMetrics(MetricRegistry metricRegistry) {
     publicAccessLogRequestRate =
         metricRegistry.meter(MetricRegistry.name(PublicAccessLogHandler.class, "RequestArrivalRate"));
     healthCheckRequestRate = metricRegistry.meter(MetricRegistry.name(HealthCheckHandler.class, "RequestArrivalRate"));
+    channelStatusInconsistent =
```
It seems like this may fit better as a counter, looking at other similar metrics (see line 107-113). However, they both should work about the same in practice, so if you prefer Meter keep it as is.

zzmao commented:
Comments addressed.
